runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/SignatureVerificationInterceptor.scala

```diff
@@ -0,0 +1,176 @@
+
+import com.daml.grpc.interceptors.ForwardingServerCallListener
+import com.google.common.io.BaseEncoding
+import io.grpc.Metadata.Key
+import io.grpc._
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.util.Try
+
+final class SignatureVerificationInterceptor(
+    keyRepository: KeyRepository.Read,
+    signedCommands: SignedCommandRepository.Write,
+) extends ServerInterceptor {
+
+  import SignatureVerificationInterceptor._
+
+  override def interceptCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      metadata: Metadata,
+      next: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+
+    val requestMetadata =
+      for {
+        signature <- getHeader(metadata, signature)
+        algorithm <- getHeader(metadata, algorithm)
+        fingerprint <- getHeader(metadata, fingerprint)
+        key <- getKey(keyRepository, fingerprint)
+      } yield (signature, algorithm, key)
+
+    requestMetadata match {
+      case Right((signature, signatureAlgorithm, signaturePublicKey)) =>
+        new SignatureVerificationServerCallListener(
+          call = call,
+          metadata = metadata,
+          next = next,
+          signature = signature,
+          signatureAlgorithm = signatureAlgorithm,
+          signaturePublicKey = signaturePublicKey,
+          signatures = signedCommands,
+        )
+      case Left(rejection) =>
+        rejection.report()
+        call.close(SignatureVerificationFailed, new Metadata())
+        new ServerCall.Listener[ReqT] {}
+    }
+
+  }
+
+}
+
+object SignatureVerificationInterceptor {
+
+  private object Rejection {
+
+    def fromException(throwable: Throwable): Rejection =
+      Failure("An exception was thrown while verifying the signature", throwable)
+
+    def missingHeader[A](header: Key[A]): Rejection =
+      Error(s"Malformed request did not contain header '${header.name}'")
+
+    def missingKey(fingerprint: String): Rejection =
+      Error(s"No key found for fingerprint $fingerprint")
+
+    val KeyVerificationFailed: Rejection =
+      Error("Key verification failed")
+
+    final case class Error(description: String) extends Rejection
+
+    final case class Failure(description: String, cause: Throwable) extends Rejection
+
+  }
+
+  private trait Rejection {
+    def report(): Unit = {
+      this match {
+        case Rejection.Error(reason) =>
+          logger.debug(reason)
+        case Rejection.Failure(reason, cause) =>
+          logger.debug(reason, cause)
+      }
+    }
+  }
+
+  val SignatureVerificationFailed: Status =
+    Status.UNAUTHENTICATED.withDescription("Signature verification failed")
+
+  private def getKey(
+      keys: KeyRepository.Read,
+      fingerprint: String,
+  ): Either[Rejection, PublicKey] = {
+    logger.trace("Retrieving key for fingerprint '{}'", fingerprint)
+    keys.get(fingerprint).toRight(Rejection.missingKey(fingerprint))
+  }
+
+  private val signature: Key[String] =
```

Comment:
worth sharing between client and server?

Reply:
Yeah, I thought about that. I guess I can create a further `non-repudiation-core` library containing the definitions for this. Does it makes sense?

Reply:
Sounds good to me

Reply:
Addressed in https://github.com/digital-asset/daml/pull/8608/commits/fa059b7bf14d62788619f7d584c2c7505554dc71